### PR TITLE
Learning Rate Finder

### PR DESCRIPTION
**Pull request by sai-prasanna**
Adds a new command `find-lr` that allows one to search for learning rate range where loss drops rapidly. This addresses feature request https://github.com/allenai/allennlp/issues/537 . Refer the following [blog post](https://medium.com/@surmenok/estimating-optimal-learning-rate-for-a-deep-neural-network-ce32f2556ce0) linked in that issue for overview of how the finder works.

The major changes are making few of the fields in `Trainer` "public" (ie remove underscore from names). I have used matplotlib to plot learning rate vs loss graph.

I haven't written unit tests, if the current code is ok, will do. I am a little unsure on what to test exactly.

### CODE REVIEW

**Comment by matt-gardner** [comment]
@sai-prasanna, thanks, we appreciate the PR!  A bunch of us have a pretty busy couple of weeks coming up as we're trying to wrap up some things for the end of the quarter, though, so it might take a little bit longer than usual to get to this.  If we haven't responded by the beginning of October, ping us again.  But in the meantime, you could work on fixing the issues that are causing the TeamCity build to fail.

**Comment by sai-prasanna** [comment]
@matt-gardner Fixed the type error and linting errors, except for the import one. 

Matplotlib requires one to call `matplotlib.use('Agg')` before importing pyplot. Any idea how to make pylint ignore the `wrong-import-position` errors caused by doing this ?

https://github.com/allenai/allennlp/blob/f9b5f92d5d52bf2dd65ae8cc1db22ab9b17acaef/allennlp/commands/find_learning_rate.py#L42

**Comment by matt-gardner** [comment]
Thanks for the reminder on this PR!  It had fallen off of my radar.  I should be able to look at this tomorrow.  To answer your question, you can do `# pylint: disable=multiple-statements,wrong-import-position`.

**Comment by sai-prasanna** [comment]
@matt-gardner  No problem. I fixed the pylint and mypy errors, but now getting documentation error.

> undocumented module: allennlp.commands.find_learning_rate

Where and what to document?

File: allennlp/commands/find_learning_rate.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+
+    trainer = Trainer.from_params(model,
+                                  serialization_dir,
+                                  iterator,
+                                  train_data,
+                                  params=trainer_params,
+                                  validation_data=None,
+                                  validation_iterator=None)
+
+    learning_rates, losses = search_learning_rate(trainer, start_lr,
```
Comment (DeNeutoy) [inline allennlp/commands/find_learning_rate.py]: Could you add a couple of logging statements here telling the user what is happening - e.g `Starting Learning Rate Search` and `Saving Learning rate plot to {some path}`

Comment (sai-prasanna) [reply allennlp/commands/find_learning_rate.py]: Will add it and the docs and ping

**Comment by sai-prasanna** [comment]
@DeNeutoy  Added logging statements and the documentation changes.

**Comment by DeNeutoy** [comment]
@sai-prasanna sorry last thing, can you test your contribution by adding a test here: https://github.com/allenai/allennlp/tree/master/allennlp/tests/commands

**Comment by sai-prasanna** [comment]
Would a test with a simple model and asserting whether the graph gets created do?

Little unsure on how to test the effects of command parameters like stopping factor etc.

**Comment by DeNeutoy** [comment]
@sai-prasanna you don't need to fix those pylint warnings in your test case - just add `pylint: disable=bad-continuation` at the top of the file. How you have them indented is fine for json in a test.

**Comment by DeNeutoy** [comment]
@sai-prasanna thanks for contributing and being so patient!

**Comment by sai-prasanna** [comment]
@DeNeutoy Thanks for your review.
